#
# This is a Shiny web application. You can run the application by clicking
# the 'Run App' button above.
#
# Find out more about building applications with Shiny here:
#
#    http://shiny.rstudio.com/
#

library(shiny)
library(shinyjs)


# Common backend scripts

source("scripts/99_backend_utils.R")
source("scripts/00_backend.R")

# Backend scripts Junction Search
source("scripts/tab1/01_backend_tab1.R")
source("scripts/tab1/02_backend_set_general_ui_data.R")
source("scripts/tab1/03_backend_visualise_in_transcript.R")
source("scripts/tab1/04_backend_download_data.R")


# Backend scripts Gene Visualisation
source("scripts/tab2/01_backend_tab2.R")

# Backend scripts Metadata
source("scripts/tab4/01_backend_tab4.R")



# Define UI for application that draws a histogram
ui <- navbarPage("IntroVerse 2.0",
                 
                 inverse=TRUE,
                 id = "introverse_menu",
                 selected = "home",
                 
                 useShinyjs(),
                 
                 tags$head(
                   tags$link(rel = "icon", type = "image/png", href = "https://raw.githubusercontent.com/SoniaRuiz/introverse/main/introverse/www/introverse.ico"),
                   tags$link(rel = "stylesheet", type = "text/css", href = "api.css"),
                   ## Google Analytics
                   # Google tag (gtag.js)
                   tags$script(src = "https://www.googletagmanager.com/gtag/js?id=G-PD5SEMENW0", "async"),
                   tags$script(src = "google_analytics.js"),
                   ## IGV
                   tags$script(src = "https://cdn.jsdelivr.net/npm/igv@2.15.5/dist/igv.min.js"),
                   tags$script(src = "igv.js")
                 ),
                 
                 tabPanel(title = "Home",
                          value = "home",
                          icon = icon("house"),
                          
                          id = "tabSwitch",
                          
                          fluidPage(
                            fluidRow(
                              column(width = 2),
                              column(width = 8, 
                                     
                                     #div(align = "center",),
                                     div(
                                       h1("About IntroVerse 2.0"),
                                       p("IntroVerse is a web resource developed with the goal of harmonizing intron-level splicing information across a wide variety of 
                                       publicly available large-scale human RNA-sequencing projects, making relevant summary data available for the wider scientific community."),
                                       p("IntroVerse 2.0 includes splicing information across >350K unique annotated introns and >6 million alternative 5' and 3' splicing events (GRCh38)
                                       across more than 27K individual RNA-sequencing human samples of diverse ancestries and cell lines, sequenced as part of various disease-specific and population RNA-sequencing studies, which include: 
                                         i) the Genotype-Tissue Expression (GTEx) dataset, covering >17K human samples from 54 non-diseased tissues, ii) The Cancer Genome Atlas Program (TGCA), spanning >11K human samples from 33 cancer types and matched control samples, and iii) ENCODE's extensive analysis of RBP activity generated by targeting >80 RBPs using either short-hairpin or CRISPR knockdowns to generate a case-control data set of >1.1K RNA-sequencing samples."),
                                       p("In addition, IntroVerse 2.0 has been enriched with information on RBP-RNA interactions supported by >20m binding sites generated through >17K CLIP-seq experiments (PMID: 24297251). 
                                         We have combined intron-level information with ClinVar data (PMID: 29165669), providing the user with granular information about the location and classification of potential pathogenic variation. 
                                         Finally, we have embedded the full functionality of the Integrative Genomics Viewer (IGV) browser as an additional feature, allowing users to visualise overlaid RNA-seq read coverage across sample groups to boost the interpretability of complex splicing data."),
                                       p("All data included on IntroVerse 2.0 originate from publicly-available datasets and it is released for the benefit of the scientific community, without restriction on use."),
                                     
                                       h2("Publications"),
                                       p("Check out these publications to learn more about IntroVerse and how to use it for research interpretation."),
                                       tags$ul(
                                         tags$li(a(href="https://academic.oup.com/nar/article/51/D1/D167/6833238", target="_blank", 
                                                   "IntroVerse: a comprehensive database of introns across human tissues, Nucleic Acids Research, Volume 51, Issue D1, 6 January 2023, Pages D167â€“D178")),
                                         tags$li(a(href="https://www.biorxiv.org/content/10.1101/2023.03.29.534370v1.full", target="_blank", "Splicing accuracy varies across human introns, tissues and age"))
                                         ),
                                       h2("Database build"),
                                       p("To build each of the 4 databases included in IntroVerse 2.0, the following strict quality-control and split read/transcript process was applied. 
                                         The major steps are summarised in the image below:"),
                                       img(src = "front_diagram.drawio.png"),
                                       hr(),
                                       br()
                                     )
                              ),
                              column(width = 2)
                            )
                          )
                 ),
                 tabPanel(title = "Junction Search",
                          
                          value="basic_search",
                          
                          icon = icon("magnifying-glass"),
                          shinybusy:: use_busy_spinner(
                            spin_id = "main_spinner",
                            spin = "circle",
                            color = "#0dc5c1",
                            position = "full-page"
                          ),
                          
                          # Sidebar with a slider input for number of bins 
                          sidebarLayout(
                            sidebarPanel(
                              width = 3,
                              p(strong("Insert the coordinates of an annotated intron, alternative 5' or alternative 3' splicing event of interest (hg38):")),
                              shiny::splitLayout(id="chr_strand_tab1",
                                                 # shiny::textInput(inputId = "jxn_coordinates_tab1",
                                                 #                  label = "Coordinates (hg38):",
                                                 #                  value = "chr16:86478133-86508654:-")
                                                 shiny::selectInput(inputId = "chr_tab1",
                                                                    label = "Chr",
                                                                    selected = 4,
                                                                    choices = c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,"X","Y"),
                                                                    multiple = F),
                                                 shiny::selectInput(inputId = "strand_tab1",
                                                                    label = "Strand",
                                                                    choices = c("+", "-"),
                                                                    selected = "-",
                                                                    multiple = F)
                              ),
                              shiny::splitLayout(id="start_end_tab1",
                                                 
                                                 numericInput(inputId = "start_tab1",
                                                              label = "Start",
                                                              value = 89835693),
                                                 numericInput(inputId = "end_tab1",
                                                              label = "End",
                                                              value = 89836742)                   
                              ),
                              hr(),
                              #shiny::splitLayout(id="splitLayout_nearby_tab1",
                              shiny::textInput(inputId = "database_text_nearby_tab1",
                                               label = "Expand the search region around the starting and ending positions indicated above (in bp):",
                                               placeholder = "Number of base-pairs",
                                               value = 0),
                              shiny::sliderInput(inputId = "database_nearby_tab1", 
                                                 label = NULL, 
                                                 step = 1,
                                                 value = 0, 
                                                 min = 0, 
                                                 max = 100000 ),
                                                 
                              #),
                              hr(),
                              p(strong("Select the database and sample group to perform the query.")),
                              
                                                 shiny::selectInput(inputId = "database_selection_tab1",
                                                                    label = "Database (RNA-seq project)",
                                                                    selected = "tcga",
                                                                    choices = c("All" = "all",
                                                                                "The Cancer Genome Atlas Program (TCGA)" = "tcga",
                                                                                "The Genotype-Tissue Expression (GTEx) v8" = "gtex",
                                                                                
                                                                                #"PD/Control" = "pd",
                                                                                #"AD/Control" = "ad"
                                                                                
                                                                                "ENCODE CRISPR" = "encode_crispr",
                                                                                "ENCODE shRNA" = "encode_sh"),
                                                                    multiple = F),
                              
                                                shiny::selectInput(inputId = "database_project_selection_tab1",
                                                                   label = "Sample experiment",
                                                                   choices = NULL,
                                                                   multiple = F),
                              
                                                 shiny::selectInput(inputId = "database_sample_type_selection_tab1",
                                                                    label = "Sample group",
                                                                    choices = NULL,
                                                                    multiple = F),
                   
                              hr(),
                              actionButton(inputId = "acceptButton_tab1", class = "btn-primary", label = "Accept"),
                              
                              ## Hidden inputs for the transcript visualisation
                              shinyjs::hidden(shiny::textInput(inputId = "nearby_helper_tab1", label = "")),
                              shinyjs::hidden(
                                p(id = "junID_tab1", ""),
                                p(id = "geneName_tab1", ""),
                                p(id = "junType_tab1", ""),
                                p(id = "transcriptENS_tab1", ""),
                                p(id = "transcriptMANE_tab1", ""),
                                p(id = "databaseName_tab1", ""),
                                p(id = "clinvarlocus_tab1", ""),
                                p(id = "phastCons17_5ss_tab1",""),
                                p(id = "phastCons17_3ss_tab1",""),
                                p(id = "CDTS_5ss_tab1",""),
                                p(id = "CDTS_3ss_tab1",""),
                                p(id = "MES_5ss_tab1",""),
                                p(id = "MES_3ss_tab1",""),
                                p(id = "donor_sequence_tab1",""),
                                p(id = "acceptor_sequence_tab1",""),
                                
                                p(id = "CLNSIG_list_tab1",""),
                                p(id = "CLNVC_list_tab1",""),
                                p(id = "MC_list_tab1","")
                              )
                              
                            ),
                            
                            # Show the output
                            mainPanel(
                              width = 9,
                              div(
                                id="welcome_help_tab1",
                                h2("Junction Search"),
                                h3(htmltools::em("Cross-database query of individual splicing events")),
                                p("Please, indicate a chromosome, start, end and strand positions corresponding to a splicing event of interest. Then, select a database (RNA-seq project) from which to extract information about the splicing event indicated."),
                                p("If the imputed coordinates precisely match with a splicing event in the database (or databases) selected, this query will return two main pieces of information."), 
                                p("The first section will contain global and unvarying information about the splicing event. For example, its length (in base-pairs), type of splicing event (i.e. annotated intron, alternative 5' or 3' splicing event), visualisation within its MANE transcript (if available), number of overlapping RBP binding sites and ClinVar variants, inter-species conservation across primates and human constraint levels of the 100-bp sequence surrounding the two splice sites, etc,."),
                                p("Secondly, it will display a tabular structure containg the specific splicing activity found across the samples of each database (in which the splicing event has been found). For example, sample frequency, total number of supporting split reads, Mis-Splicing Ratio measures if the splicing event is an annotated intron, median TPM value of the gene from which the splicing event originated, etc.")
                              ),
                              shiny::uiOutput(outputId = "queryResults"), 
                              shiny::uiOutput(outputId = "queryNearByResults"),
                              tabsetPanel(
                                id = "switcher",
                                type = "tabs" 
                              ),
                              
                              shinyBS::bsModal(id = "modalVisualiseTranscript_tab1", 
                                               title = "Visualise the splicing event within the selected transcript",
                                               trigger = NULL,
                                               size = "large",
                                               
                                               shinybusy:: use_busy_spinner(
                                                 spin_id = "modal_spinner",
                                                 spin = "circle",
                                                 color = "#0dc5c1",
                                                 position = "full-page"
                                               ),
                                               
                                               plotOutput("modalVisualiseTranscript_tab1", width = "auto", height = "70vh")),
                              
                              shinyBS::bsModal(id = "modalVisualiseCLIP_tab1", 
                                               title = "Visualising top RBP-RNA regulatory data supported by the highest number of CLIP-seq experiments.",
                                               
                                               trigger = NULL,
                                               size = "large",
                                               
                                               shinybusy:: use_busy_spinner(
                                                 spin_id = "modal_spinner",
                                                 spin = "circle",
                                                 color = "#0dc5c1",
                                                 position = "full-page"
                                               ),
                                               
                                               
                                               plotOutput("modalVisualiseCLIP_tab1", width = "auto", height = "75vh"),
                                               footer = downloadButton(outputId = 'downloadCLIPData', label = 'Download all regulatory data', class = "btn-primary")
                                               ),
                              
                              shinyBS::bsModal(id = "modalVisualiseClinVar_tab1", 
                                               title = "ClinVar pathogenic/likely-pathogenic splicing variants",
                                               
                                               trigger = NULL,
                                               size = "large",
                                               shinybusy:: use_busy_spinner(
                                                 spin_id = "modal_spinner",
                                                 spin = "circle",
                                                 color = "#0dc5c1",
                                                 position = "full-page"
                                               ),
                                               
                                               uiOutput("modalVisualiseClinVar_tab1", width = "auto", height = "65vh")),
                              
                              shinyBS::bsModal(id = "modalVisualiseMORE_tab1", 
                                               title = "More information about the splicing event selected",
                                               trigger = NULL,
                                               size = "large",
                                               uiOutput(outputId = "modalVisualiseMORE_tab1", width = "auto", height = "70vh") )
                             
                            )
                          )
                          ),
                 tabPanel(title = "Gene Visualisation",
                          value = "gene_search",
                          icon = icon("dna"),
                          
                          shinybusy:: use_busy_spinner(
                            spin_id = "tab2_spinner",
                            spin = "circle",
                            color = "#0dc5c1",
                            position = "full-page"
                          ),
                          
                          
                          sidebarLayout(
                            sidebarPanel(
                              width = 3,
                              p(strong("Please, select the database and preferred visualisation mode to represent alternative 3' and 5' splicing events:")),
                              
                              shiny::selectInput(inputId = "database_tab2",
                                                 label = "Database (RNA-seq project)",
                                                 selected = "tcga",
                                                 choices = c("The Cancer Genome Atlas Program (TCGA)" = "tcga",
                                                             "The Genotype-Tissue Expression (GTEx) v8" = "gtex",
                                                             
                                                             #"PD/Control" = "pd",
                                                             #"AD/Control" = "ad",
                                                            
                                                             "ENCODE CRISPR" = "encode_crispr",
                                                             "ENCODE shRNA" = "encode_sh"),
                                                 selectize=T,
                                                 multiple = F),
                              
                              shiny::selectInput(inputId = "database_project_tab2",
                                                 label = "Sample experiment",
                                                 choices = NULL,
                                                 multiple = F, selectize=T),
                              
                              shiny::selectInput(inputId = "database_sample_type_tab2",
                                                 label = "Sample group",
                                                 choices = NULL,
                                                 multiple = F),
                              shiny::selectInput(inputId = "splicing_event_type_tab2",
                                                 label = "Splicing event type:",
                                                 choices = c("All junction types" = "all",
                                                             "Annotated intron" = "intron", 
                                                             "Novel acceptor (alternative 3'ss)" = "novel_acceptor",
                                                             "Novel donor (alternative 5'ss)" = "novel_donor"),
                                                 selected = "all",
                                                 multiple = F),
                              shiny::selectInput(inputId = "visualise_mode_tab2",
                                                 label = "Visualisation mode:",
                                                 choices = NULL,
                                                 selected = "all",
                                                 multiple = F),
                              hr(),
                              p(strong("Please, select the gene and transcript structure of interest to visualise splicing events:")),
                              selectizeInput(inputId = "gene_tab2",
                                             label = "Gene:",
                                             choices = NULL,
                                             multiple = F,
                                             options = list(
                                               placeholder = "Select gene",
                                               options = list(create = FALSE)),
                                             selected = NULL),
                              
                              selectizeInput(inputId = "transcript_tab2",
                                             label = "Transcript (Ensembl v111):",
                                             choices = NULL,
                                             multiple = F,
                                             options = list(
                                               placeholder = "Select transcript",
                                               options = list(create = FALSE)),
                                             selected = NULL),
                              hr(),
                              actionButton(inputId = "acceptButton_tab2", class = "btn-primary", label = "Visualise Splicing")
                            ),
                            mainPanel(
                              width = 9,
                              div(
                                id="welcome_help_tab2",
                                h2("Gene visualisation"),
                                h3(htmltools::em("Visualisation of all splicing events detected within a transcript structure")),
                                p("This feature provides a visual overview of all splicing events detected across the introns of a given transcript structure and gene of interest across the samples of the database (RNA-seq project) selected."),
                                p("To facilitate data interpretation, splicing events have been coloured by junction type (i.e. annotated introns in blue, alternative 5' and 3' splicing events in green and purple, respectively) and labelled with their total number of supporting split reads across the samples of the database and sample experiment selected. This information will be additionally provided within a data table structure, which can be ordered, filtered and downloaded in multiple file formats.")
                              ),
                              shiny::uiOutput(outputId = "geneQueryResults")
                              )
                          )
                          ),
                 tabPanel(title = "IGV Browser",
                          value = "igv_browser",
                          icon = icon("window-maximize"),
                          fluidPage(
                            fluidRow(
                              column(width = 12, 
                                     h2("Interactive Genome Browser (IGV)"),
                                     h3(htmltools::em("Overlaid BigWig Visualisation across samples")),
                                     p("This feature embeds the full functionality of the Integrative Genomics Viewer (IGV) browser (PMID: 36562559). It provides interactive coverage visualisation and comparison through the visualisation of overlaid RNA-seq read coverage across sample groups. This feature has been included to complement and boost the interpretation of individual splicing events of interest.")
                                    # p("This section allows the visualisation of BigWig files. Please, select the sample experiment of interest. To establish comparisons among sample groups, please select 'All' in the sample type dropbox. Given the non-disease nature of the GTEx samples, and the large amount of samples provided by tissue (in some cases >700 samples), a maximum of 50 samples per tissue have been selected for visualisation purposes. To facilitate comparisons across samples groups in the TGCA dataset, only matching control and disease samples from the same individual have been selected.")
                              ),
                              column(width = 12, 
                                     shiny::splitLayout(id="igv_browser_dataset_selection",
                                                        
                                                        shiny::selectInput(inputId = "igv_browser_database_selection_tab1",
                                                                           label = "Database (RNA-seq project)",
                                                                           selected = "tcga",
                                                                           choices = c("The Cancer Genome Atlas Program (TCGA)" = "tcga",
                                                                                       "The Genotype-Tissue Expression (GTEx) v8" = "gtex",
                                                                                       
                                                                                       #"AD/Control" = "ad",
                                                                                       #"PD/Control" = "pd",
                                                                                       
                                                                                       "ENCODE CRISPR" = "encode_crispr",
                                                                                       "ENCODE shRNA" = "encode_sh"
                                                                           ),
                                                                           width = '100%',
                                                                           multiple = F),
                                                        shiny::selectInput(inputId = "igv_browser_database_project_selection_tab1",
                                                                           label = "Sample cluster",
                                                                           choices = NULL,
                                                                           width = '100%',
                                                                           multiple = F),
                                                        shiny::selectInput(inputId = "igv_browser_database_sample_type_selection_tab1",
                                                                           label = "Sample group",
                                                                           choices = NULL,
                                                                           width = '100%',
                                                                           multiple = F),
                                                        br()
                                     ),
                                     shiny::splitLayout(id="igv_browser_button",
                                                        actionButton(inputId = "igv_browser_acceptButton_tab1", class = "btn-primary", 
                                                                     
                                                                     label = "Visualise BigWig"),
                                                        shinyjs::hidden(
                                                          p(id = "bigwig_urls", value=""),
                                                          p(id = "bigwig_categories", value=""),
                                                          p(id = "bigwig_types", value="")
                                                        ))
                              )
                            ),
                            fluidRow(
                              column(width = 12, 
                                     hr(),
                                     div(id = "igv-div", align = "center")
                                     
                              )
                            )
                          )
                 ),
                 tabPanel("Datasets",
                          icon = icon("database"),
                          shinybusy:: use_busy_spinner(
                            spin_id = "main_spinner_tab2",
                            spin = "circle",
                            color = "#0dc5c1",
                            position = "full-page"
                          ),
                          
                          fluidPage(
                            
                            navlistPanel(
                              id = "tab_panel_DB_info",
                              widths = c(3, 9),
                              "Bulk Short-Read RNA-seq:",
                              
                              tabPanel(
                                title = "TCGA",
                                fluidRow(
                                  column(width = 12,
                                         h1("TCGA"),
                                         h3(strong("The Cancer Genome Atlas Program")),
                                         p("The Cancer Genome Atlas (TCGA) is a landmark cancer genomics program, molecularly characterized over 20,000 primary cancer and matched normal samples spanning 33 cancer types"),
                                         plotOutput("tcga_metadata_output_tab2", width = "auto", height = "60vh"),
                                         
                                         uiOutput(outputId = "tcga_cancer_types_output_tab2")
                                  ))),
                              tabPanel(
                                title = "GTEx v8",
                                fluidRow(
                                  column(width = 12,
                                         h1("GTEx v8"),
                                         h3(strong("Genotype Tissue Expression Project v8")),
                                         p("GTEx project collected samples from up to 54 non-diseased tissue sites across nearly 1,000 deceased individuals. Gene expression of each tissue was assessed by RNA sequencing (bulk RNA-seq)."),
                                         plotOutput("gtex_metadata_output_tab2", width = "auto", height = "60vh")
                                  ))),
                              tabPanel(
                                title = "ENCODE shRNA",
                                
                                fluidRow(
                                  column(width = 12,
                                         h1("ENCODE shRNA"),
                                         h3(strong("shRNA knockdown followed by RNA-seq and control experiments")),
                                         p("Bulk short-read RNA-seq on K562 and HepG2 cell lines treated with an shRNA knockdown against different RNA-Binding Proteins (RBPs) as well as against no target gene (controls). BAM files were downloaded from the ENCODE platform."),
                                         plotOutput("ENCODE_shRNA_metadata_output_tab2", width = "auto", height = "60vh")
                                  ))),
                              tabPanel(
                                title = "ENCODE CRISPR",
                                
                                fluidRow(
                                  column(width = 12,
                                         h1("ENCODE CRISPR"),
                                         h3(strong("Homo sapiens K562 and HepG2 genetically modified (deletion) using CRISPR targeting multiple RNA-Binding Proteins (RBPs) and control experiments")),
                                         p("Bulk short-read RNA-seq on K562 and HepG2 cell lines genetically modified (deletion) using CRISPR against 71 independent RNA-Binding Proteins (RBPs) as well as against no target gene (controls). BAM files were downloaded from the ENCODE platform."),
                                         plotOutput("ENCODE_crispr_metadata_output_tab2", width = "auto", height = "60vh")
                                  )))
                              
                            ))
                 )
)

# Define server logic required to draw a histogram
server <- function(input, output, session) {
  
  #updateSelectizeInput(session, 'gtex_table_list', choices = gtex_table_list, server = TRUE, selected = gtex_table_list[1])
  #updateSelectizeInput(session, 'encode_table_list', choices = encode_table_list, server = TRUE, selected = encode_table_list[1])
  
  
  # shiny::updateTabsetPanel(inputId = "introverse_menu", selected = "Basic Search")
  
  
  ##############################################################################
  ## OBSERVERS - JUNCTION SEARCH TAB
  ##############################################################################
 
  
  # Manage the selection of the database and project dropdowns
  observeEvent(list(input$database_selection_tab1), {
                      
    ## Database Selection - we update the dropdown "sample cluster selection"
    if (input$database_selection_tab1 == "all") {
      
      ## Tab Basic Search
      updateSelectInput(session, inputId = "database_project_selection_tab1", choices = c("All" = "all"))
      updateSelectInput(session, inputId = "database_sample_type_selection_tab1", choices = c("All" = "all"))
      
      shinyjs::disable(id = "database_project_selection_tab1")
      shinyjs::disable(id = "database_sample_type_selection_tab1")
      
    
    } else {
      
      ## Tab Basic Search
      sample_project_choices <- set_database_project_dropdown(database_key = input$database_selection_tab1)
      updateSelectInput(session, inputId = "database_project_selection_tab1", choices = sample_project_choices, selected = sample_project_choices[2])
      shinyjs::enable(id = "database_project_selection_tab1")
 
    }
    
  }, ignoreInit = F)
  
  # Manage the selection of the project and sample cluster dropdowns
  observeEvent(input$database_project_selection_tab1, {
  
    message(input$database_selection_tab1, input$database_project_selection_tab1 )

    ## Sample Cluster Selection - we update the dropdown "sample type selection"
    if ( input$database_selection_tab1 == "all" || input$database_project_selection_tab1 == "all" ) {

      updateSelectInput(session, inputId = "database_sample_type_selection_tab1", choices = c("All" = "all"))
      shinyjs::disable(id = "database_sample_type_selection_tab1")

    } else {

      ## Basic Search
      sample_cluster_choices <- set_database_sample_type_dropdown(database_key = input$database_selection_tab1, 
                                                                  project_id = input$database_project_selection_tab1)
      if (sample_cluster_choices %>% length() == 1) {
        updateSelectInput(session, inputId = "database_sample_type_selection_tab1", choices = sample_cluster_choices)
      } else {
        updateSelectInput(session, inputId = "database_sample_type_selection_tab1", choices = sample_cluster_choices,
                          selected = sample_cluster_choices[2])
      }
      shinyjs::enable(id = "database_sample_type_selection_tab1")
      
    }

  }, ignoreInit = T)
  
  
  # When the value of any of the following inputs is changed, we update them visually
  # Start position
  observeEvent(input$start_tab1,{
    session$sendInputMessage("start_tab1", list(value=input$start_tab1))
  })
  # End position
  observeEvent(input$end_tab1,{
    session$sendInputMessage("end_tab1", list(value=input$end_tab1))
  })
  # Genomic region slider
  observeEvent(input$database_nearby_tab1,{
    session$sendInputMessage("database_nearby_tab1", list(value=input$database_nearby_tab1))
  })
  observeEvent(input$database_text_nearby_tab1, {
    updateSliderInput(session, inputId = "database_nearby_tab1", 
                      value = input$database_text_nearby_tab1)
  })
  observeEvent(input$nearby_helper_tab1,{
    session$sendInputMessage("nearby_helper_tab1", list(value=input$nearby_helper_tab1))
  })
  
  
  ##############################################################################
  ## OBSERVERS - GENE VISUALISATION TAB
  ##############################################################################
  
  
  observeEvent(input$database_tab2, {

    ## 1. The list of genes available  may change depending on the database selected
    chosen_gene_list <- gene_list[database_equivalences %>% filter(key == input$database_tab2) %>% pull(sqlite_file)][[1]]
    genes_choices <- c(chosen_gene_list$gene_name, chosen_gene_list$gene_id) %>% as.list()
    names(genes_choices) <- c(chosen_gene_list$gene_name, chosen_gene_list$gene_id) %>% as.list()
    updateSelectizeInput(session, 'gene_tab2', choices = genes_choices, server = TRUE)  
    
    ## 2. Update the list of projects available depending on the database selected
    sample_project_choices <- set_database_project_dropdown(database_key = input$database_tab2)
    if (sample_project_choices[1] == "all") {
      sample_project_choices <- sample_project_choices[-1]
    }
    updateSelectInput(session, inputId = "database_project_tab2", choices = sample_project_choices, selected = sample_project_choices[1])

  }, ignoreInit = F)
  
  observeEvent(input$database_project_tab2, {
    
    sample_cluster_choices <- set_database_sample_type_dropdown(database_key = input$database_tab2, project_id = input$database_project_tab2)
    
    if ( sample_cluster_choices %>% length() == 1) {
      updateSelectInput(session, inputId = "database_sample_type_tab2", choices = sample_cluster_choices)
      updateSelectInput(session, inputId = "visualise_mode_tab2", choices = c("All splicing events" = F))
      
    } else {
      updateSelectInput(session, inputId = "database_sample_type_tab2", choices = sample_cluster_choices[-1], selected = sample_cluster_choices[2])
      updateSelectInput(session, inputId = "visualise_mode_tab2", choices = c("All splicing events" = F,
                                                                              "Unique splicing events to this sample group" = T))
    }

  }, ignoreInit = T)
  
  observeEvent(input$gene_tab2, {
    
    shinyjs::disable(id = "acceptButton_tab2")
    
    chosen_conn <- conn_list[database_equivalences %>% filter(key == input$database_tab2) %>% pull(sqlite_file)][[1]]
    transcripts_gene = get_database_transcripts(con = chosen_conn, gene_name = input$gene_tab2)
    
    transcript_choices <- c(transcripts_gene$transcript_id) %>% as.list()
    names(transcript_choices) <- c(transcripts_gene$transcript_label) %>% as.list()
    updateSelectizeInput(session, 'transcript_tab2', choices = transcript_choices, server = TRUE)  
    shinyjs::enable(id = "acceptButton_tab2")
  }, ignoreInit = T)
  
  
  ##############################################################################
  ## OBSERVERS - IGV BROWSER
  ##############################################################################
  
  
  # Manage the selection of the database and project dropdowns
  observeEvent(list(input$igv_browser_database_selection_tab1), {
    
    sample_project_choices <- set_database_project_dropdown(database_key = input$igv_browser_database_selection_tab1)
    if (sample_project_choices[1] == "all") {
      sample_project_choices <- sample_project_choices[-1]
    }
    ## Tab IGV Browser
    updateSelectInput(session, 
                      inputId = "igv_browser_database_project_selection_tab1", 
                      choices = sample_project_choices, 
                      selected = sample_project_choices[1])
    
    shinyjs::enable(id = "igv_browser_database_project_selection_tab1")
    
  }, ignoreInit = F)
  
  # Manage the selection of the project and sample cluster dropdowns
  observeEvent(list(input$igv_browser_database_project_selection_tab1), {
    
    sample_cluster_choices <- set_database_sample_type_dropdown(database_key = input$igv_browser_database_selection_tab1, 
                                                                project_id = input$igv_browser_database_project_selection_tab1)
    
    ## IGV Browser
    if (sample_cluster_choices %>% length() == 1) {
      updateSelectInput(session, inputId = "igv_browser_database_sample_type_selection_tab1", choices = sample_cluster_choices)
    } else {
      updateSelectInput(session, inputId = "igv_browser_database_sample_type_selection_tab1", choices = sample_cluster_choices,
                        selected = sample_cluster_choices[1])
    }
    shinyjs::enable(id = "igv_browser_database_sample_type_selection_tab1")
    
  }, ignoreInit = T)
  
  # Manage the selection of the correct bigWig files
  observeEvent(input$igv_browser_acceptButton_tab1, {
    
    database_path <- file.path(here::here(), "database/bigwig.sqlite")
    
    ## Load data
    tryCatch(
      eval(con <- DBI::dbConnect(RSQLite::SQLite(), database_path)),
      error = function(e) {
        showNotification(paste("Error in 'dbConnect': ", e$message), type = "error")
      }
    )
    
    bigWig <- DBI::dbGetQuery(con, paste0("SELECT * FROM 'bigwig'")) %>% as_tibble()
    DBI::dbDisconnect(conn = con)
    
    message(input$igv_browser_database_project_selection_tab1 , " ",
            input$igv_browser_database_sample_type_selection_tab1 %>% str_to_title())
    
    bigWig <- bigWig %>%
      filter(project == input$igv_browser_database_project_selection_tab1 )
    
    
    ## Retrieve the BigWig URLs
    if (input$igv_browser_database_sample_type_selection_tab1 != "all") {
      
    
      bigWig <- bigWig %>%
        filter(cluster == input$igv_browser_database_sample_type_selection_tab1 %>% str_to_title())
    }

    tryCatch(
      eval(shinyjs::html(id = "bigwig_categories", html = paste(bigWig$cluster, collapse = ";"))),
      error = function(e) {
        showNotification(paste("Error in 'bigwig_categories': ", e$message), type = "error")
      }
    )
    tryCatch(
      eval(shinyjs::html(id = "bigwig_urls", html = paste(bigWig$BigWigURL, collapse = ";"))),
      error = function(e) {
        showNotification(paste("Error in 'bigwig_urls': ", e$message), type = "error")
      }
    )
    tryCatch(
      eval(shinyjs::html(id = "bigwig_types", html = paste(bigWig$BigWig_type, collapse = ";"))),
      error = function(e) {
        showNotification(paste("Error in 'bigwig_types': ", e$message), type = "error")
      }
    )
    
  
    shinyjs::runjs(code = "loadBigWigFiles($('#bigwig_categories').html(), $('#bigwig_urls').html(), $('#bigwig_types').html());")
    
  }, ignoreInit = T)
  
  
  ##############################################################################
  ## MODAL POP-UP 
  ##############################################################################
  
  ## TRANSCRIPT VISUALISATION ----------------------------------------------------
  
  visualiseTranscriptPlot <- function() {
    
    if (input$transcriptMANE_tab1) {
      transcript_plot <- visualise_transcript(junID = str_replace_all(string = input$junID_tab1, pattern = "%20", replacement = " "),
                                              transcript_ENS = str_replace_all(string = input$transcriptENS_tab1, pattern = "%20", replacement = " "),
                                              database_name = str_replace_all(string = input$databaseName_tab1, pattern = "%20", replacement = " "),
                                              junType = str_replace_all(string = input$junType_tab1, pattern = "%20", replacement = " "))
    } else {
      transcript_plot <- visualise_multiple_transcripts(junID = str_replace_all(string = input$junID_tab1, pattern = "%20", replacement = " "),
                                                        transcript_ENS = str_replace_all(string = input$transcriptENS_tab1, pattern = "%20", replacement = " "),
                                                        geneName = str_replace_all(string = input$geneName_tab1, pattern = "%20", replacement = " "),
                                                        database_name = str_replace_all(string = input$databaseName_tab1, pattern = "%20", replacement = " "),
                                                        junType = str_replace_all(string = input$junType_tab1, pattern = "%20", replacement = " "))
    }

    shinybusy::hide_spinner(spin_id = "modal_spinner")
    return(transcript_plot)
  }
  
  output$modalVisualiseTranscript_tab1 <- renderPlot({
    shinybusy::show_spinner(spin_id = "modal_spinner")
    shinyjs::runjs(code = '$("#modalVisualiseTranscript_tab1 > img").addClass("d-none");')
    shinyjs::runjs(code = '$("#modalVisualiseTranscript_tab1 > img").remove();')
    visualiseTranscriptPlot()
  }, res = 75)
  
  
  
  ## CLIP VISUALISATION ---------------------------------------------------------
  
  visualiseCLIP_Plot <- function() {
    CLIP_plot <- visualise_CLIP(junID = str_replace_all(string = input$junID_tab1, pattern = "%20", replacement = " "),
                                #transcript_ENS = str_replace_all(string = input$transcriptENS_tab1, pattern = "%20", replacement = " "),
                                geneName = str_replace_all(string = input$geneName_tab1, pattern = "%20", replacement = " "),
                                junType = str_replace_all(string = input$junType_tab1, pattern = "%20", replacement = " "))
    shinybusy::hide_spinner(spin_id = "modal_spinner")
    return(CLIP_plot)
  }
  
  output$modalVisualiseCLIP_tab1 <- renderPlot({
    shinybusy::show_spinner(spin_id = "modal_spinner")
    shinyjs::runjs(code = '$("#modalVisualiseCLIP_tab1 > img").addClass("d-none");')
    shinyjs::runjs(code = '$("#modalVisualiseCLIP_tab1 > img").remove();')
    visualiseCLIP_Plot()
  }, res = 75)
  
  
  ## CLINVAR VISUALISATION ----------------------------------------------------
  
  visualiseClinVarPlot <- function() {
    clinvar_plot <- visualise_clinvar(junID = str_replace_all(string = input$junID_tab1, pattern = "%20", replacement = " "),
                                      #clinvar_locus = str_replace_all(string = input$clinvarlocus_tab1, pattern = "%20", replacement = " "),
                                      
                                      #CLNSIG_list = str_replace_all(string = input$CLNSIG_list_tab1, pattern = "%20", replacement = " "),
                                      #CLNVC_list = str_replace_all(string = input$CLNVC_list_tab1, pattern = "%20", replacement = " "),
                                      #MC_list = str_replace_all(string = input$MC_list_tab1, pattern = "%20", replacement = " "),
                                      
                                      geneName = str_replace_all(string = input$geneName_tab1, pattern = "%20", replacement = " "),
                                      database_name = str_replace_all(string = input$databaseName_tab1, pattern = "%20", replacement = " "),
                                      junType = str_replace_all(string = input$junType_tab1, pattern = "%20", replacement = " "))
    shinybusy::hide_spinner(spin_id = "modal_spinner")
    return(clinvar_plot)
  }
  
  output$modalVisualiseClinVar_tab1 <- renderUI({
    shinybusy::show_spinner(spin_id = "modal_spinner")
    shinyjs::runjs(code = '$("#modalVisualiseClinVar_tab1 > img").addClass("d-none");')
    shinyjs::runjs(code = '$("#modalVisualiseClinVar_tab1 > img").remove();')
    
    tagList(
      
      renderPlot({
        visualiseClinVarPlot()
      }),
      br(),
      DT::renderDataTable(expr = DT::datatable(data = get_database_clinvar(get_genomic_coordinates(str_replace_all(string = input$junID_tab1, pattern = "%20", replacement = " "))),
                                               rownames = F, 
                                               extensions = c('Buttons', 'Responsive'),
                                                                         options = list(
                                                                           pageLength = 10,
                                                                           dom = 'Bfrtip',
                                                                           buttons = c('copy', 'csv', 'excel')
                                                                           ))
      )
    )
    
    
    
  })#, res = 72)
  
  
  ## 'MORE...' VISUALISATION ----------------------------------------------------
  
  visualiseMORE_Table <- function() {
    
    DT::datatable(data = data.frame("Property" = c("PhastCons17 5'ss (100 bp)","PhastCons17 3'ss (100 bp)","CDTS 5'ss (100 bp)",
                                                   "CDTS 3'ss (100 bp)","MaxEntScan 5'ss (9 bp)","MaxEntScan 3'ss (23 bp)", 
                                                   "5'ss sequence (3bp | 6bp)", "3'ss sequence (20bp | 3bp)" ),
                                    "Value" = c(input$phastCons17_5ss_tab1,
                                                input$phastCons17_3ss_tab1,
                                                input$CDTS_5ss_tab1,
                                                input$CDTS_3ss_tab1,
                                                input$MES_5ss_tab1,
                                                input$MES_3ss_tab1,
                                                input$donor_sequence_tab1,
                                                input$acceptor_sequence_tab1),
                                    "Description" = c("Inter-species conservation across 17 primates. Calculated using the 100 bp sequence overlapping the 5' splice site downstream the intron. 
                                                      The higher is the PhastCons17 score, the more conserved is the 100 bp sequence evaluated across primates.",
                                                      
                                                      "Inter-species conservation across 17 primates. Calculated using the 100 bp sequence overlapping the 3' splice site upstream the intron.",
                                                      
                                                      "Sequence constraint across humans. Calculated using the 100 bp sequence overlapping the 5' splice site downstream the intron.\n
                                                      The higher is the CDTS score, the higher sequence variation across humans for the sequence evaluated.",
                                                      
                                                      "Sequence constraint across humans of the 100 bp sequence overlapping the 3' splice site into the upstream intron.",
                                                      
                                                      "Sequence motif similarity of the 9 bp sequence overlapping the 5'ss splice sequence to known 5' splice sites in the human reference annotation. 
                                                      The higher is the score, the higher is the likelihood of the 5' splice site being recognised by the spliceosome machinery.",
                                                      
                                                      "Sequence motif similarity of the 23 bp sequence overlapping the 3'ss splice sequence to known 3' splice sites in the human reference annotation.", 
                                                      
                                                      "9 bp motif DNA sequences located at the 5'ss (-3|+6 bp) of the junction evaluated (Human Primary DNA Assembly hg38).",
                                                      "23 bp motif DNA sequences located at the 3'ss (-20|+3 bp) of the junction evaluated (Human Primary DNA Assembly hg38)." )),
                  rownames = F, 
                  extensions = c('Buttons', 'Responsive'),
                  
                  options = list(
                    pageLength = 10,
                    dom = 'Bfrtip',
                    buttons = c('copy', 'csv', 'excel')
                  )) %>%
      return()
  }
  
  output$modalVisualiseMORE_tab1 <- renderUI({
    shinyjs::runjs(code = '$("#modalVisualiseMORE_tab1 > img").addClass("d-none");')
    shinyjs::runjs(code = '$("#modalVisualiseMORE_tab1 > img").remove();')
    visualiseMORE_Table() %>% return()
  })

  
  ##############################################################################
  ## JUNCTION SEARCH - RETRIEVE SPLICING INFO FROM DATABASES 
  ## TAB #1
  ##############################################################################

  
  toListenNearBy <- eventReactive(input$acceptButton_tab1, {
    
    shinybusy::show_spinner(spin_id = "main_spinner")
    
    query_results <- query_database_nearby(chr = input$chr_tab1,
                                           start = input$start_tab1,
                                           end = input$end_tab1,
                                           strand = input$strand_tab1,
                                           databases_name = input$database_selection_tab1,
                                           project_id = input$database_project_selection_tab1,
                                           table_name = input$database_sample_type_selection_tab1,
                                           nearby_window = input$database_nearby_tab1)
    
    shinybusy::hide_spinner(spin_id = "main_spinner")
  
    div(
      h1("Splicing events found in the genomic region ",
         paste0("[chr",input$chr_tab1, ":", 
                (input$start_tab1 %>% as.integer() - input$database_nearby_tab1 %>% as.integer()), " - ", 
                (input$end_tab1 %>% as.integer() + input$database_nearby_tab1 %>% as.integer()), ":",
                input$strand_tab1,"]")),
      br(),
      h4(strong("This region has been calculated after expanding the upstream and downstream search window by ", input$database_nearby_tab1, " base pairs.")),
      p("All splicing events that overlap with the indicated genomic region across the selected database, sample cluster and sample type are shown below."),
      br(),
      query_results %>%
        DT::renderDataTable(server = F,
                            extensions = c('Buttons', 'Responsive'),
                      options = list(order = list(0, 'desc'),
                                     pageLength = -1,
                                     dom = 'Bfrtip',
                                     buttons = c('copy', 'csv', 'excel')),
                      rownames = F,
                      callback = DT::JS("table.rows().every(function(i, tab, row) {
                      
                      var $this = $(this.node());
                      var nearby_value = document.getElementById('database_nearby_tab1').value;

                      // We store the value of the slider range in a hidden text field before resetting it back again to zero value
                      var onclick_f = 'Shiny.setInputValue(\"nearby_helper_tab1\",' + nearby_value + ');Shiny.setInputValue(\"database_nearby_tab1\",' + 0 + ');Shiny.setInputValue(\"start_tab1\",' + this.data()[2] + ');Shiny.setInputValue(\"end_tab1\",' + this.data()[3] + ');Shiny.onInputChange(\"acceptButton_tab1\",' + i + ',{priority:\"event\"});';
                      
                      var num = '<a id=\"' + i + '\" role=\"button\" onclick = ' + onclick_f + '> ' + this.data()[0] + '</a>';
                      
                      $this.find('td:eq(0)', row).html(num);
                      
                      });"))
      ) %>%
      return()
  })
  
  output$queryNearByResults <- renderUI({
    
    req(input$acceptButton_tab1, cancelOutput = F)
    
    ## Remove previous tabs
    for (tab_name in database_equivalences$name){
      shiny::removeTab(inputId = "switcher", target = tab_name)  
    }
    
    ## Hide previous results
    shinyjs::hide(id = "switcher")
    shinyjs::hide(selector = ".tabbable div.tab-content")
    
    ## Query database
    if ( isolate(input$database_nearby_tab1) > 0 ) {
      toListenNearBy()  
    }
    
  })
  
  toListen <- eventReactive(input$acceptButton_tab1, {

    shinybusy::show_spinner(spin_id = "main_spinner")
    
    ## Query databases
    query_results <- query_database(chr = input$chr_tab1,
                                    start = input$start_tab1,
                                    end = input$end_tab1,
                                    strand = input$strand_tab1,
                                    databases_name = input$database_selection_tab1,
                                    project_id = input$database_project_selection_tab1,
                                    table_name = input$database_sample_type_selection_tab1)

    
    if ( query_results %>% nrow > 0 ) {

      
      
      #message(shinyjs::runjs("alert($('#nearby_helper_tab1').val());alert( document.getElementById('nearby_helper_tab1').value);"))
      
      ## Info about the junction that is common to all databases
      v <- setup_UI_details_section(query_results, back_button = input$nearby_helper_tab1)
      
      i <- (v %>% length()) + 1
      v[[i]] <- div(
        br(),
        hr(),
        h3("Splicing information per database"),
        br()
      )
      
      ## Create a tab per database containg the splicing info particular to each database
      result_tabs <- setup_UI_results_section(query_results)
      
      if ( result_tabs %>% length > 0 ) {
        
        lapply(1:length(result_tabs), FUN =  function(j) {
          shiny::insertTab(
            inputId = "switcher",
            result_tabs[[j]],
            target = NULL,
            select = T
          )
        })
        
        ## Show new tab results
        shinyjs::show(id = "switcher")
        shinyjs::show(selector = ".tabbable div.tab-content")
      } 
      
      ## Hide loading spinner
      shinybusy::hide_spinner(spin_id = "main_spinner")
      
      ## Return results
      return(v)
      
    } else {
      shinybusy::hide_spinner(spin_id = "main_spinner")
      p(strong("No results found using the criteria selected!"))
    }
    
  })

  output$queryResults <- renderUI({

    req(input$acceptButton_tab1, cancelOutput = F)

    ## Remove previous tabs
    for (tab_name in database_equivalences$name){
      shiny::removeTab(inputId = "switcher", target = tab_name)  
    }
    
    ## Hide previous results
    shinyjs::hide(id = "welcome_help_tab1")
    shinyjs::hide(id = "switcher")
    shinyjs::hide(selector = ".tabbable div.tab-content")
    
    ## Query database
    if ( isolate(input$database_nearby_tab1) == 0 ) {
      toListen()
    }
    
  })
  
  
  
  ##############################################################################
  ## GENE SEARCH - RETRIEVE SPLICING INFO FROM DATABASES 
  ## TAB #2
  ##############################################################################
  
  toListenGene <- eventReactive(input$acceptButton_tab2, {
    
    shinybusy::show_spinner(spin_id = "tab2_spinner")
    
    ## Get the database to query
    database_sqlite <- database_equivalences %>% filter(key == input$database_tab2) %>% pull(sqlite_file)
    
    
    # message(input$gene_tab2, " ", input$transcript_tab2, " ", database_sqlite, " ", input$database_project_tab2, " ", input$database_sample_type_tab2, " ", 
    #         input$visualise_mode_tab2, " ", input$splicing_event_type_tab2)
    
    gene_data_to_visualise <- get_gene_splicing_data_to_visualise(gene.id = input$gene_tab2,
                                                                  transcript.id = input$transcript_tab2,
                                                                  database.sqlite = database_sqlite,
                                                                  project.id = input$database_project_tab2,
                                                                  table.name = input$database_sample_type_tab2,
                                                                  junction.type = input$splicing_event_type_tab2,
                                                                  compare = input$visualise_mode_tab2)

    shinybusy::hide_spinner(spin_id = "tab2_spinner")

    if ( all(names(gene_data_to_visualise) == "error") ) {

      tagList(

        shiny::h1(database_equivalences %>% filter(sqlite_file == database_sqlite) %>% pull(name_title)),
        br(),
        renderPlot({
          visualise_empty_plot(text = gene_data_to_visualise$error)
        })
      )

    } else {

      
      tagList(

        shiny::h1(database_equivalences %>% filter(sqlite_file == database_sqlite) %>% pull(name_title)),
        br(),


        #database_sqlite <- databases_sqlite_list[1]

        #lapply(1:(data_gene_comparison$table_name %>% length()), function(n) {    ## Query visualisation gene splicing in selected database

          # n <- 1

        
        renderPlot({
          visualise_gene_data(gene_splicing_data = gene_data_to_visualise)
        }, width = "auto", height = 600),

        #shiny::h3("Details:"),
        DT::renderDataTable(expr = DT::datatable(data = gene_data_to_visualise$junctions_to_plot  %>%
                                                   mutate(junction_type = str_replace_all(junction_type, pattern = "_", replacement = " ") %>% str_to_title()) %>%
                                                   dplyr::select(-c(ID,junction_color,intron_color,u2_intron)) %>%
                                                   
                                                   dplyr::rename("Seqnames" = seqnames,
                                                                 "Start" = start,
                                                                 "End" = end,
                                                                 "Strand" = strand,
                                                                 "Transcript ID" = transcript_id,
                                                                 "Total split read counts" = sum_counts,
                                                                 "Junction type" = junction_type),
                                                 rownames = T,
                                                 extensions = c('Buttons', 'Responsive'),
                                                 caption = htmltools::tags$caption(
                                                   style = 'caption-side: top; color:black;',
                                                   htmltools::h2('Junction details:')),
                                                 options = list(
                                                   order = list(list(8, 'asc')), #list(2, 'asc')),
                                                   pageLength = 15,
                                                   dom = 'Bfrtip',
                                                   buttons = c('copy', 'csv', 'excel')
                                                 )) ,
                            server = T),
        shinybusy::hide_spinner(spin_id = "tab2_spinner")



        #}),
        
      )
      
    }
    
  })
  
  output$geneQueryResults <- renderUI({
    
    req(input$acceptButton_tab2, cancelOutput = F)
    
    shinyjs::hide(id = "welcome_help_tab2")
    
    toListenGene()
    
    
  })
  
  ##################################################
  ## DATABASES INFO TAB - RETRIEVE METADATA OUTPUT
  ##################################################
  
  toListen_getDBMetadata <- eventReactive(list(input$tab_panel_DB_info), {
    shinybusy::show_spinner(spin_id = "main_spinner_tab2")
    data_to_plot <- visualise_metadata_sample_clusters(input$tab_panel_DB_info)  
    shinybusy::hide_spinner(spin_id = "main_spinner_tab2")
    
    data_to_plot %>% return()
  })
  
  output$tcga_metadata_output_tab2 <- renderPlot({
    toListen_getDBMetadata()
  })
  
  output$gtex_metadata_output_tab2 <- renderPlot({
    toListen_getDBMetadata()
  })
  
  output$ENCODE_shRNA_metadata_output_tab2 <- renderPlot({
    toListen_getDBMetadata()
  })
  
  output$ENCODE_crispr_metadata_output_tab2 <- renderPlot({
    toListen_getDBMetadata()
  })
  
  output$tcga_cancer_types_output_tab2 <- renderUI({
    div(h3(strong("Cancer types")),
        DT::datatable(data = table_cancer_types(),
                      rownames = F)) %>% return()
  })
  
  ###########################################
  ## DOWNLOAD HANDLERS
  ###########################################

  output$downloadCLIPData <- downloadHandler(
    filename = function() {
      paste("CLIPdata_",input$junID_tab1, "_", input$geneName_tab1,".csv", collapse = "")
    },
    content = function(file) {
      CLIP_data <- download_CLIP_data(input$junID_tab1, input$geneName_tab1)
      write.csv(x = CLIP_data, 
                file, 
                row.names = F,
                fileEncoding = "UTF-8")
    }
  )
  
}

# Run the application 
shinyApp(ui = ui, server = server)

