# IntroVerse 2.0 Update: a comprehensive database of introns across human tissues and disease

## About IntroVerse 2.0

IntroVerse 2.0 is a web resource developed with the goal of harmonizing intron-level splicing information across a wide variety of publicly available large-scale human RNA-sequencing projects, making relevant summary data available for the wider scientific community.

IntroVerse 2.0 includes splicing information across >350K unique annotated introns and >6 million alternative 5' and 3' splicing events (GRCh38) across more than 27K individual RNA-sequencing human samples of diverse ancestries and cell lines, sequenced as part of various large-scale publicly-available RNA-sequencing studies, which include: i) the Genotype-Tissue Expression (GTEx) dataset, covering >17K human samples from 54 non-diseased tissues in a tissue- and age-stratification manner (PMID: 32913098); ii) The Cancer Genome Atlas Program (TGCA), spanning >11K human samples from 33 cancer types and matched control samples (PMID: 24071849); and iii) ENCODE's extensive analysis of RBP activity generated by targeting >140 RBPs using either short-hairpin or CRISPR knockdowns to generate a case-control data set of >1.1K RNA-sequencing samples (PMID: 32728246).

In addition, IntroVerse 2.0 has been enriched with information on i) RBP-RNA interactions supported by >20m binding sites generated through >17K CLIP-seq experiments (PMID: 24297251) and ii) ClinVar data (PMID: 29165669), the latter providing with granular information about the location and classification of potential pathogenic variation. Finally, we have embedded the full functionality of the Integrative Genomics Viewer (IGV) browser as an additional feature, allowing users to visualise overlaid RNA-seq read coverage across sample groups to boost the interpretability of complex splicing data.

All data included on IntroVerse 2.0 originate from publicly-available datasets and it is released for the benefit of the scientific community, without restriction on use.

## Website

IntroVerse 2.0 is freely available for use and data download at [https://rytenlab.com/browser/app/introverse_v2](https://rytenlab.com/browser/app/introverse_v2).

## Related Publications

Check out these publications to learn more about IntroVerse 2.0 and how to use it for research interpretation.

* IntroVerse: a comprehensive database of introns across human tissues S García-Ruiz et. al, Nucleic Acids Research, Volume 51, Issue D1, 6 January 2023, Pages D167–D178, [https://doi.org/10.1093/nar/gkac1056](https://doi.org/10.1093/nar/gkac1056).
* Splicing accuracy varies across human introns, tissues and age S García-Ruiz et. al, bioRxiv 2023.03.29.534370; doi: [https://doi.org/10.1101/2023.03.29.534370](https://doi.org/10.1101/2023.03.29.534370).

## Docker Image

The dockerised version of IntroVerse 2.0 is available at [https://hub.docker.com/r/soniaruiz/introverse_v2](https://hub.docker.com/r/soniaruiz/introverse_v2).
